migrate/20240510_add_tainted_at_vm.rb

```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    alter_table(:vm) do
+      add_column :tainted_at, :timestamptz
```

Comment:
tainted typically means something became bad/unusable. And maybe also that it should be destroyed. Also, I'm wondering if we should distinguish between the time we take a vm from a pool and the time it becomes available to a customer. Potentially something like this:

`allocated_from_pool_at`:  the time we took the vm from the pool
`supplied_at`: the time it became available to the user

Reply:
I borrowed the term 'tainted' from @fdr, who suggested to mark virtual machines. We can use more positive work. `tainted` might have little negative impact.

The time we take a vm from a pool and the time it becomes available to a customer have distinct implications for different resources. For the standard VM product, these two times are same. However, for the runner product, we use the `ready_at` column to differentiate the latter. Thus, the final product should include a column to distinguish this. If we consider the runner service as a customer for the standard VM service, we have a time when the VM becomes available to the runner (which is the same as when it is picked from the pool), and when the runner becomes available to the end customer.

Reply:
yeah, fair, taint is often with some kind of badness, though per computer science tradition badness can be pretty abstract, e.g. in garbage collectors. In our case, the badness is increasing restriction on how we can manipulate the VM, as it could plausibly have data that we absolutely must not leak between customers, and not delete in a casual manner. In this respect it has some similarity how the jargon is used in garbage collectors (or even virtual memory, "dirty pages" is a related concept), where objects are often "tainted" as somehow being more complex to collect.

We could keep this wording in that tradition or have a new one.

As an aside, something I did previously was the inverse, and not at all for wording reasons, which was more like a "purity" (or, to get outside that nomenclature, a "flexibility") flag that objects started with and then was removed the moment it was plausible a customer might have put their irreplaceable data in it.

This took the form of a `can_recreate` semaphore that relevant objects started with and would be decr'd at some point during their provisioning process, when it was plausible irreplaceable user data might have entered the system. If `recreate` was incremented and `can_recreate` was still present, then the object could be re-provisioned. This "double-trigger" style provided very precise results in not permitting mistakes, and being concurrency safe. I'm not saying it's necessarily superior to what this might do, it's more just for reflection by the author.

I suppose the nice thing about that system is that the overwhelmingly common case results in `can_recreate` disappearing, i.e. there is zero bloat in the common case, provided one things that knowing the far-distant `tainted_at` time is extraneous...which it may not be

Reply:
thinking it over, how many times would we reference this symbol? maybe three, no more than six times? it is okay to avoid a jargon at expense of length. `has_customer_data` or something like that might work.

Reply:
I'm personally ok if you decide to go with "tainted". Just wanted to bring this up.

Reply:
IMO, both are fine, but I'd prefer something similar to `has_customer_data`, which requires less thinking about its function if you don't have the context.